### ISSUE TITLE
Using healthcheck on swarm disturbs nameservices

### ISSUE THREAD

**Issue by PascalTurbo** [moby/moby#35451]
After enabling healthcheck in our setup we weren't able to deploy this setup anymore. We've found out, that when enabling healthcheck containers can't resolve their service names. This includes the name of the container itself. 

We are deploying via `docker stack deploy --compose-file docker-compose.yml minitest` and tested it using docker 17.03, 17.06 and 17.10. We could observe the problem on all three versions.

I've created a minimal example. The two containers are pinging each other and as a primitive healthcheck we ping localhost.

Using docker-compose with this compose-file will lead to a working setup; using stack deploy will restart the containers time and time again.

```
version: '3'

services:
  mini-1:
    image: debian
    deploy:
      replicas: 1
    command: bash -c "sleep 15; ping mini-2 > /tmp/mini-1.log"
    healthcheck:
      test: ["CMD", "ping", "-c", "1", "localhost"]
      interval: 1m30s
      timeout: 10s
      retries: 3
    volumes:
      - /minimal_example/mini-1.log:/tmp/mini-1.log      
    networks:
      mini-network:
        aliases:
          - mini

  mini-2:
    image: debian
    deploy:
      replicas: 1
    command: bash -c "sleep 15; ping mini-1 > /tmp/mini-2.log"
    healthcheck:
      test: ["CMD", "ping", "-c", "1", "localhost"]
      interval: 1m30s
      timeout: 10s
      retries: 3
    volumes:
      - /minimal_example/mini-2.log:/tmp/mini-2.log
    networks:
      mini-network:
        aliases:
          - mini

networks:
  mini-network:
    driver: overlay
```

**Comment by thaJeztah** [comment]
In your stack you're pinging service "mini-2" from "mini-1", so the "ping" won't succeed if the other service is not yet running. As a result, the command fails, and the container exits. If that happens, docker will start a new container ("task") to replace it, and the issue will occur again.

You can see the same happening with a `docker run` below:


```bash
$ docker run -d --name bla debian bash -c "sleep 15; ping mini-2 > /tmp/mini-1.log"
e5f0ce1e678c403857935c5dd7cacfb26846d8a11abe7785cca1fd2ffc9c12a8
```

After 15 seconds:

```bash
$ docker ps -a

CONTAINER ID        IMAGE                                 COMMAND                  CREATED                  STATUS                      PORTS                    NAMES
e5f0ce1e678c        debian                                "bash -c 'sleep 15; …"   Less than a second ago   Exited (1) 13 seconds ago                            bla
```

The _healthcheck_ determins if a service is enabled on the network or not; if the service is not healthy, it's not advertised on the network, and no traffic is routed to it, which would explain why the name of the service is not discoverable in DNS.

**Comment by PascalTurbo** [comment]
@thaJeztah thank you for your answer. You could also try the example with pinging the local service instead of the remote one.
But maybe your last sentence will answer why we have these problems with our setup.

**Comment by edmundkwok** [comment]
I'm experiencing this as well, tested on 17.12.0, 17.12.1 and the latest edge 18.05.1. I can replicate it consistently with MongoDB with a replica set, using the official `healthcheck/mongo` that only adds an additional HEALTCHECK command with the official `mongo` image as a base. Steps to reproduce:

1. Initialize a 3 member MongoDB replica set in a Docker Swarm - respective hostnames: `mongo1`, `mongo2`, `mongo3`
2. Once all replica members are healthy (checked with `rs.status()` in mongo console), run `docker stop` on a secondary, in this example, `mongo2`.
3. When the `mongo2` replica restarts, it will fail to rejoin the replica set. `docker logs` indicates that the local hostname `mongo2` cannot resolve.

`mongo2` Logs:

```
2018-05-21T05:40:31.456+0000 W NETWORK  [replexec-0] getaddrinfo("mongo2") failed: Name or service not known
2018-05-21T05:40:31.458+0000 I NETWORK  [replexec-0] getaddrinfo("mongo2") failed: Name or service not known
2018-05-21T05:40:31.479+0000 W REPL     [replexec-0] Locally stored replica set configuration does not have a valid entry for the current node; waiting for reconfig or remote heartbeat; Got "NodeNotFound: No host described in new configuration 43061556 for replica set rs0 maps to this node" while validating { _id: "rs0", version: 43061556, protocolVersion: 1, members: [ { _id: 0, host: "mongo1:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 1.0, tags: {}, slaveDelay: 0, votes: 1 }, { _id: 1, host: "mongo2:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 0.5, tags: {}, slaveDelay: 0, votes: 1 }, { _id: 2, host: "mongo3:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 0.5, tags: {}, slaveDelay: 0, votes: 1 }, { _id: 3, host: "mongo4:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 0.0, tags: {}, slaveDelay: 0, votes: 1 } ], settings: { chainingAllowed: true, heartbeatIntervalMillis: 2000, heartbeatTimeoutSecs: 10, electionTimeoutMillis: 10000, catchUpTimeoutMillis: -1, catchUpTakeoverDelayMillis: 30000, getLastErrorModes: {}, getLastErrorDefaults: { w: 1, wtimeout: 0 }, replicaSetId: ObjectId('595a2325441b58d6a497b001') } }
2018-05-21T05:40:31.479+0000 I REPL     [replexec-0] New replica set config in use: { _id: "rs0", version: 43061556, protocolVersion: 1, members: [ { _id: 0, host: "mongo1:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 1.0, tags: {}, slaveDelay: 0, votes: 1 }, { _id: 1, host: "mongo2:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 0.5, tags: {}, slaveDelay: 0, votes: 1 }, { _id: 2, host: "mongo3:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 0.5, tags: {}, slaveDelay: 0, votes: 1 }, { _id: 3, host: "mongo4:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 0.0, tags: {}, slaveDelay: 0, votes: 1 } ], settings: { chainingAllowed: true, heartbeatIntervalMillis: 2000, heartbeatTimeoutSecs: 10, electionTimeoutMillis: 10000, catchUpTimeoutMillis: -1, catchUpTakeoverDelayMillis: 30000, getLastErrorModes: {}, getLastErrorDefaults: { w: 1, wtimeout: 0 }, replicaSetId: ObjectId('595a2325441b58d6a497b001') } }
```

4. `mongo2` can only rejoin if we force it with `rs.recofig(rs.conf(), {force: true})` in mongo console
5. If we now update `mongo2` to use the official `mongo` image, it restarts and rejoins the replica set automatically. I can confirm that I can also replicate this building my own Docker Mongo image with (fails to rejoin) and without (successfully rejoins) the `HEALTHCHECK` command.

`docker info`:

```
Containers: 9
 Running: 5
 Paused: 0
 Stopped: 4
Images: 19
Server Version: 18.05.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: iwe4487qva0b6hg6haf8wp6de
 Is Manager: false
 Node Address: 10.99.0.5
 Manager Addresses:
  10.99.0.12:2377
  10.99.0.13:2377
  10.99.0.15:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-109-generic
Operating System: Ubuntu 16.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.953GiB
Name: db-1-2
ID: ZGVO:FQ7D:UURC:RXNO:AMNG:YVUG:ETFT:UCFF:HWQN:Q42Y:5RMF:D4PJ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
 provider=vultr
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

`docker version`:

```
Client:
 Version:      18.05.0-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   f150324
 Built:        Wed May  9 22:16:25 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.05.0-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   f150324
  Built:        Wed May  9 22:14:32 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Comment by thaJeztah** [comment]
> that only adds an additional `HEALTHCHECK` command with the official mongo image as a base

Just to prevent confusion; a docker container / image can only have a single `HEALTHCHECK`. Looks like the official mongo image doesn't have a health check defined, so I guess you meant "adds a HEALTHCHECK" not an "additional one" 😅 

> When the `mongo2` replica restarts, it will fail to rejoin the replica set. docker logs indicates that the local hostname `mongo2` cannot resolve.

Looks like a chicken-and-egg situation; your health check wants to connect to the `mongo2` host to connect to itself (the container), but the `mongo2` host (DNS entry) won't become available until the container becomes healthy.

Instead of attempting to connect to `mongo2`, could you make the health check connect to `localhost`?

**Comment by edmundkwok** [comment]
Thanks for your response @thaJeztah!

Sorry for the confusion, my choice of words was not the best 😅

As for the `HEALTHCHECK` itself, please allow me to clarify. The issue is not the healtcheck itself that is not able to resolve `mongo2`, but MongoDB itself is complaining on startup that it cannot resolve `mongo2` (as per the logs attached earlier).

The HEALTHCHECK itself is running fine, and it does fallback to `127.0.0.1` if the hostname is not available, as per:

https://github.com/docker-library/healthcheck/blob/master/mongo/docker-healthcheck#L4

Just to add more context to the issue (replicated in Play With Docker):

1. Deploy MongoDB (`docker stack deploy -c docker-compose.yml mongo`) with the following Compose file:

```
version: "3.4"
services:
  mongo1:
    image: healthcheck/mongo
    command: mongod --replSet "rs0" --bind_ip 0.0.0.0
    networks:
      - mongo-network
    volumes:
      - mongo1:/data/db
    deploy:
      placement:
        constraints:
          - node.role == manager
  mongo2:
    image: healthcheck/mongo
    command: mongod --replSet "rs0" --bind_ip 0.0.0.0
    networks:
      - mongo-network
    volumes:
      - mongo2:/data/db
    deploy:
      placement:
        constraints:
          - node.role == manager
  mongo3:
    image: healthcheck/mongo
    command: mongod --replSet "rs0" --bind_ip 0.0.0.0
    networks:
      - mongo-network
    volumes:
      - mongo3:/data/db
    deploy:
      placement:
        constraints:
          - node.role == manager

networks:
  mongo-network:

volumes:
  mongo1:
  mongo2:
  mongo3:
```
2. Initialize MongoDB replica set as per https://docs.mongodb.com/manual/tutorial/deploy-replica-set/. i.e. start a Mongo shell and run the following command:

```
rs.initiate({
   _id : "rs0",
   members: [
      { _id: 0, host: "mongo1:27017" },
      { _id: 1, host: "mongo2:27017" },
      { _id: 2, host: "mongo3:27017" }
   ]
})
```

3. Mongo `rs.status()` confirms that there are 1 Primary (`mongo1`) and 2 Secondary (`mongo2` && `mongo3`).
4. Kill `mongo2` and wait for Docker Swarm to restart
5. `mongo2` is marked as healthy in `docker ps` as the healthcheck only pings the mongo service instead of the replica set status. To confirm the hostname problem, when starting a mongo shell in `mongo2`, the replica status `REMOVED`, as per `rs.status()`:

```
rs0:OTHER> rs.status()
{
        "state" : 10,
        "stateStr" : "REMOVED",
        "uptime" : 278,
        "optime" : {
                "ts" : Timestamp(1527601489, 5),
                "t" : NumberLong(1)
        },
        "optimeDate" : ISODate("2018-05-29T13:44:49Z"),
        "ok" : 0,
        "errmsg" : "Our replica set config is invalid or we are not a member of it",
        "code" : 93,
        "codeName" : "InvalidReplicaSetConfig",
        "operationTime" : Timestamp(1527601489, 5),
        "$clusterTime" : {
                "clusterTime" : Timestamp(1527601986, 1),
                "signature" : {
                        "hash" : BinData(0,"AAAAAAAAAAAAAAAAAAAAAAAAAAA="),
                        "keyId" : NumberLong(0)
                }
        }
}
```

6. `docker logs <mongo2_container_id>` contains the following lines which stands out
**NETWORK  [replexec-0] getaddrinfo("mongo2") failed: Name or service not known**.

MongoDB logs in context:

```
2018-05-29T13:48:32.566+0000 W NETWORK  [replexec-0] getaddrinfo("mongo2") failed: Name or service not known
2018-05-29T13:48:32.571+0000 I NETWORK  [replexec-0] getaddrinfo("mongo2") failed: Name or service not known
2018-05-29T13:48:32.573+0000 W REPL     [replexec-0] Locally stored replica set configuration does not have a valid entry for the current node; waiting for reconfig or remote heartbeat; Got "NodeNotFound: No host described in new configuration 1 for replica set rs0 maps to this node" while validating { _id: "rs0", version: 1, protocolVersion: 1, members: [ { _id: 0, host: "mongo1:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 1.0, tags: {}, slaveDelay: 0, votes: 1 }, { _id: 1, host: "mongo2:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 1.0, tags: {}, slaveDelay: 0, votes: 1 }, { _id: 2, host: "mongo3:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 1.0, tags: {}, slaveDelay: 0, votes: 1 } ], settings: { chainingAllowed: true, heartbeatIntervalMillis: 2000, heartbeatTimeoutSecs: 10, electionTimeoutMillis: 10000, catchUpTimeoutMillis: -1, catchUpTakeoverDelayMillis: 30000, getLastErrorModes: {}, getLastErrorDefaults: {w: 1, wtimeout: 0 }, replicaSetId: ObjectId('5b0d59445a2313982c92fb02') } }
2018-05-29T13:48:32.573+0000 I REPL     [replexec-0] New replica set config in use: { _id: "rs0", version: 1, protocolVersion: 1, members: [ { _id: 0, host: "mongo1:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 1.0, tags: {}, slaveDelay: 0, votes: 1 }, { _id: 1, host: "mongo2:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 1.0, tags: {}, slaveDelay: 0, votes: 1 }, { _id: 2, host: "mongo3:27017", arbiterOnly: false, buildIndexes: true, hidden: false, priority: 1.0, tags: {}, slaveDelay: 0, votes: 1 } ],settings: { chainingAllowed: true, heartbeatIntervalMillis: 2000, heartbeatTimeoutSecs: 10, electionTimeoutMillis: 10000, catchUpTimeoutMillis: -1, catchUpTakeoverDelayMillis: 30000, getLastErrorModes: {}, getLastErrorDefaults: { w: 1, wtimeout: 0 }, replicaSetId: ObjectId('5b0d59445a2313982c92fb02') } }
2018-05-29T13:48:32.573+0000 I REPL     [replexec-0] This node is not a member of the config
```
7. Run `docker service update --image mongo mongo_mongo2` (which is without the `HEALTHCHECK` command), and once started, it rejoins the replica set as per expected. No error about `mongo2` hostname appears in `docker logs`

**Comment by hezor** [comment]
I'm having the exact same chicken-and-egg situation with a 3-node Elasticsearch cluster. The internal DNS name of the service (in this case: elasticsearch) is not exposed until the healthcheck succeeds (curl localhost:9200), but because the DNS name cannot be used, the Elasticsearch cluster will never be started and port 9200 will never be listened.

**Comment by thaJeztah** [comment]
> container's `/etc/hosts` is overridden in our setup so it's not an option

@ushuz is there a specific reason to override `/etc/hosts` in your setup? Would it be an option to use the `--host` option to add custom entries to `/etc/hosts`?

> The internal DNS name of the service (in this case: elasticsearch) is not exposed until the healthcheck succeeds (curl localhost:9200), but because the DNS name cannot be used, the Elasticsearch cluster will never be started and port 9200 will never be listened.

@hezor Does it need to just have the DNS name _resolving_, or does it make use of the IP address that it resolves to?

Both; did some quick testing; not sure if any of the below are useful for your specific situation, but thought I'd post them here just in case

#### 1. Creating service without extra options

This should roughly match the current situation (although without "health check");

```bash
docker network create --driver=overlay --scope=swarm testnet
docker service create --name=test --network=testnet nginx:alpine

docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
b2c07d9c8c13        nginx:alpine        "nginx -g 'daemon of…"   8 seconds ago       Up 6 seconds        80/tcp              test.1.xwt94dg7rkllfrm8fveqn7cht

docker exec -it  test.1.xwt94dg7rkllfrm8fveqn7cht sh

/ # cat /etc/hosts
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
10.0.0.4	b2c07d9c8c13

/ # hostname
b2c07d9c8c13

/ # ping test
PING test (10.0.0.3): 56 data bytes
64 bytes from 10.0.0.3: seq=0 ttl=64 time=0.058 ms
64 bytes from 10.0.0.3: seq=1 ttl=64 time=0.059 ms
^C
--- test ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.058/0.058/0.059 ms
```

In this case, `test` (servicename) would only resolve once the service is marked "healthy", and resolves to the service's VIP

> **Note** I notice that the containername is not set as hostname, only the container-id. I'm not sure if this is by design, but I'll open a separate issue for that



#### 2. Creating a service with a healthcheck

This matches the situation where the service becomes healthy after some time; until the service is healthy, the servicename won't resolve;


```
docker service create \
  --detach \
  --name=test \
  --health-cmd='if [ ! -f "/count" ] ; then ctr=0; else ctr=`cat /count`; fi; ctr=`expr ${ctr} + 1`; echo "${ctr}" > /count; if [ "$ctr" -gt 3 ] ; then exit 0; else exit 1; fi' \
  --health-interval=10s \
  --health-start-period=60s \
  --network=testnet \
   nginx:alpine

docker ps

CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                                     PORTS               NAMES
eb97441b5b89        nginx:alpine        "nginx -g 'daemon of…"   3 seconds ago       Up Less than a second (health: starting)   80/tcp              test.1.9d50jotfpw7fhis7ew0csasog


docker exec -it test.1.9d50jotfpw7fhis7ew0csasog sh

/ # cat /etc/hosts
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
10.0.0.24	eb97441b5b89
/ # hostname
eb97441b5b89
/ # ping test
ping: bad address 'test'
/ # ping test
ping: bad address 'test'
...
...
/ # ping test
PING test (10.0.0.23): 56 data bytes
64 bytes from 10.0.0.23: seq=0 ttl=64 time=0.059 ms
64 bytes from 10.0.0.23: seq=1 ttl=64 time=0.063 ms
^C
--- test ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.059/0.061/0.063 ms
```


#### 3. Set each container's hostname to the service-name

While this makes the servicename resolve inside the contaienr, this may not be what you want if you want the servicename to actually resolve to the service's VIP. Setting  the hostname will cause the servicename to be resolved to the _containers_ (task) IP-address;

```bash
docker service create --name test --hostname='{{.Service.Name}}' --network=testnet nginx:alpine

/ # cat /etc/hosts
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
10.0.0.6	test

/ # hostname
test

/ # ping test
PING test (10.0.0.6): 56 data bytes
64 bytes from 10.0.0.6: seq=0 ttl=64 time=0.094 ms
64 bytes from 10.0.0.6: seq=1 ttl=64 time=0.061 ms
^C
--- test ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.061/0.077/0.094 ms
```

#### 4. Add a hosts entry for the servicename

This would also make the servicename "resolve" inside the container, but point to localhost, which may not what you want;

```bash
docker service create --name test --host=test:127.0.0.1 --network=testnet nginx:alpine

docker exec -it test.1.z3e3b2mjoodxzagnhffz6imvv sh 

/ # cat /etc/hosts
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
127.0.0.1	test
10.0.0.10	b8af29ab4b86

/ # hostname
b8af29ab4b86

/ # ping test
PING test (127.0.0.1): 56 data bytes
64 bytes from 127.0.0.1: seq=0 ttl=64 time=0.051 ms
64 bytes from 127.0.0.1: seq=1 ttl=64 time=0.076 ms
^C
--- test ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.051/0.063/0.076 ms
```

**Comment by ushuz** [comment]
> > container's /etc/hosts is overridden in our setup so it's not an option
>
> @ushuz is there a specific reason to override /etc/hosts in your setup? Would it be an option to use the --host option to add custom entries to /etc/hosts?

@thaJeztah We use `/etc/hosts` as internal name service, instead of deploying an internal DNS server. It consists of plenty of entires, and frequently changes. So it's not very doable with `--host` option.

**Comment by flybyray** [comment]
Because of that chicken egg problem some projects go with a workarround on the healthcheck itself
For instance see the healthcheck for mariadb galeradb docker swarm:
```
https://github.com/colinmollenhour/mariadb-galera-swarm/blob/master/healthcheck.sh
```
it checks a `pre-boot.flag` file which is created inside of the container if a special environment variable is set:
https://github.com/colinmollenhour/mariadb-galera-swarm/blob/master/start.sh#L324

An entrypoint script waits for the "chicken" In a startup loop https://github.com/colinmollenhour/mariadb-galera-swarm/blob/master/start.sh#L287-L340
and places the "egg" by removing `pre-boot.flag` file:
https://github.com/colinmollenhour/mariadb-galera-swarm/blob/master/start.sh#L342

The healthcheck will then go in its normal mode of reporting health.

It is somehow tricky but a stable way to get this chicken-egg problem solved within swarm - wihtout using an additional layer of complexity with some extra service discovery tool like consul or the like.

**Comment by flybyray** [comment]
you may notice the extra wait time introduced in this workarround by this while loop.

That is just because:

* docker swarm service discovery is associated with healthcheck !!

For faster startup times of cluster services (mongo,cassandra,elasticsearch,mariadb) with healthchecks enabled some additional discovery (not associated with healthchecks) needs to be integrated.

**Comment by drnybble** [comment]
@thaJeztah It seems to me that having a service name resolve through DNS only when at least one replica is healthy is a design mistake. Software expects that configured hostnames are resolvable irrespective of whether the endpoint is available at that precise moment. As can be seen by the several examples above, the way Swarm treats hostname resolution causes friction. I have seen the same problem where a Mongo node in a replicaSet starts up, can't resolve its own hostname, and considers that it is no longer part of the replicaSet! Normally you connect to the endpoint to see if it is alive -- what is wrong with doing the same within a Swarm? The hostname should ALWAYS resolve to the VIP address -- if the service exists I should be able to resolve its IP (even if it is currently scaled to 0 replicas). Also, the hostname lookup is often cached, so trying to use hostname resolution as a means of seeing if there is currently a healthy endpoint is not realistic.

**Comment by ElRakiti** [comment]
Is there any advice on how to solve that issue without changing the images? We have the same problem with initiating an elasticsearch cluster. When healthcheck is included, the initial start up fails because the hostnames are not yet available but necessary for the cluster startup. 

I totally agree with @drnybble, the hostname should be resolvable without having the service in the container fully running.

**Comment by drnybble** [comment]
Here is what I am doing (read it & weep): 
- create the services with 0 replicas
- extract the VIP from the service definition (docker service inspect...)
```
	local NETWORK_ID=$(docker network inspect --format='{{.Id}}' "${NETWORK_NAME}")

	local ADDR=$(docker service inspect --format="{{range .Endpoint.VirtualIPs}}{{if eq .NetworkID \"${NETWORK_ID}\"}}{{.Addr}}{{end}}{{end}}" "${SERVICE_NAME}")
	# ADDR=10.0.21.11/24; Trim off the /24 to get just the address
	VIP=${ADDR%/*}
```
- define environment variables with these values
- use these in the YAML file extra_hosts
- now change the replicas to what you want & update the services (replicas can be defined using environment variables)
So your service is something like this:
```
replicas: ${REPLICAS_MYSERVICE}
extra_hosts:
- "hostname:${hostname_vip}
```

**Comment by drnybble** [comment]
Actually, constructing a hosts file and supplying it as a docker config might be easier I am going to try that.

**Comment by information-security** [comment]
I am totally against the decision of bounding service name resolution to success of health-checks. Many services are healthy **only and only if** they can communicate with other services. Just as simple as that.
Please let me know if I am misunderstanding the concept here.

/cc @thaJeztah

**Comment by jerrac** [comment]
Is the fact that healthchecks have to pass before dns is set up documented anywhere? I just spent a good 10+ hours trying to figure out why I could find dns info in one project but not in another. Eventually I randomly commented out my compose file's healthcheck, and suddenly I could query for my services task ips. 

I don't see anything in the healthcheck docs [here](https://docs.docker.com/compose/compose-file/#healthcheck) or [here](https://docs.docker.com/engine/reference/builder/#healthcheck). 

I'll also throw my hat in the ring of people who don't think this is very helpful. DNS has nothing to do with whether a service is up or not. DNS is how you point your stuff where they can find out if something is working or not.

**Comment by jerrac** [comment]
I just ran into Docker's late initiation of dns causing problems again. This time I'm not even in swarm mode or using a health check. 

I have Varnish running as a proxy container in front of a php/apache app. Because Varnish requires dns to exist for it's backends before it starts, I need my php/apache service dns entry to be live for Varnish. And, of course, my varnish service wants to start faster than my php/apache service.

If I wasn't going to be pushing this to Swarm, I'd just use depends_on, but https://github.com/moby/moby/issues/31333 makes that impossible.

Anyway, I'm off to do more awkward bash scripting to work around this. Someday it would be nice if this problem was fixed.

(P. S. Not that I'm ungrateful for the work everyone contributing more than issues has done. Seriously, Docker is awesome and I'm super thankful it exists. :) )

**Comment by IsmailMarmoush** [comment]
Hi, 

I faced the same issue, took me hours to realize healthcheck needed to be green so that service is discovered, which kinda make sense as well.

Anyway I created this script for cassandra but the concept can be used for other services indeed.

The ideas are:

1. To have number of (retries X interval) grace period.
2. To have separate health checks one for single node and another when there's supposedly cluster.

This cluster health is because seed nodes sometimes do no connect together when started in parallel, so after the grace period they should be restarted if the cluster nodetool up nodes was under quorum size.

Note sure if I should put the script in here, but here's a gist
https://gist.github.com/IsmailMarmoush/fd21ce255786a4139d2a4c8ee0617235

**Comment by kryptt** [comment]
One simple option is to add a field allowing a service to start as healthy and delay posterior healthchecks.

This way, initialization and dns resolution is unchanged. Instead you have services with a grace period before they can be deemed unhealthy. This of course means they are in an undefined state until the grace period expires and health checks actually start. 

this would add an option like:
 `grace-period`: specifies the amount of time to consider the service healthy before starting health checks. This can be useful if your container needs to be registered on the network before it actually starts up. By default, Docker will not consider a service healthy until a healthcheck CMD has passed correctly (0 sec grade period), but you can use the grace-period option to consider the service healthy for purposes of discovery. Setting a grace-period automatically assures start-period is >= grace-period, 

this means once the grace-period is over the service may jump into 'starting' or any other relevant state the existing logic would dictate. 

This is maximally backwards compatible and not too complicated to implement

**Comment by SheepReaper** [comment]
Just came across this issue. I was trying to add a health check to my Portainer agent service in a stack that had experienced the same, or perhaps a similar, issue. When it's a new stack, as the containers start up, they detect they are in a swarm, but getting the IP of other tasks via `tasks.agent` at 127.0.0.11 returns "no such host". I had assumed this was a potainer-specific race condition, since, of course, not a single agent has started, hence no IPs. Perhaps this is a bug: the name should resolve to an empty list instead of "name not found," but that's not how DNS works, at least not a compliant one... Not sure where to go here, I want the healthcheck, but I need at least the first container to start ignoring the healthcheck, but the container won't start if DNS doesn't work
